code/espurna/garland.cpp

```diff
@@ -0,0 +1,223 @@
+/*
+GARLAND MODULE
+Copyright (C) 2020 by Dmitry Blinov <dblinov76 at gmail dot com>
+
+Tested on 60 led strip. 
+!!! For more leds can cause WDT rebot. Need to be carefully tested for more than 60 leds !!!
+The most time consuming operation is actually showing leds by Adafruit Neopixel. It take about 1870 mcs.
+More long strip can take more time to show. 
+Currently animation calculation, brightness calculation/transition and showing makes in one loop cycle.
+Debug output shows timings. Overal timing should be not more that 3000 ms.
+
+For longer strips have sense to separate show operation from animation/brightness calculation. Or even
+divide showing for parts and process them in different cycles.
+*/
+
+#include "garland.h"
+
+#if GARLAND_SUPPORT
+
+#include <Adafruit_NeoPixel.h>
+
+#include <vector>
+
+#include "garland/anims.h"
+#include "garland/color.h"
+#include "garland/palette.h"
+#include "garland/scene.h"
+#include "web.h"
+#include "ws.h"
+
+const char* NAME_GARLAND_ENABLED        = "garlandEnabled";
+const char* NAME_GARLAND_BRIGHTNESS     = "garlandBrightness";
+
+const char* NAME_GARLAND_SWITCH         = "garland_switch";
+const char* NAME_GARLAND_SET_BRIGHTNESS = "garland_set_brightness";
+
+constexpr int LEDS                      = GARLAND_LEDS;
+constexpr int PIN                       = GARLAND_D_PIN; // WS2812 pin number
+
+#define EFFECT_UPDATE_INTERVAL_MIN      5000  // 5 sec
+#define EFFECT_UPDATE_INTERVAL_MAX      10000 // 10 sec
+
+bool _garland_enabled                   = true;
+unsigned long _last_update              = 0;
+unsigned long _interval_effect_update;
+
+int paletteInd;
+int animInd                             = 0; 
+
+// Palette should
+std::vector<Palette> pals = {
```

Comment:
Do you mean define pals like this: Palette pals[] = { ..

I obtained next errors:

espurna\garland.cpp: In function 'void garlandLoop()':
espurna\garland.cpp:185:73: error: request for member 'size' in 'pals', which is of non-class type 'Palette [11]'
         while (prevPalInd == paletteInd) paletteInd = secureRandom(pals.size());
                                                                         ^
espurna\garland.cpp: In function 'void garlandSetup()':
espurna\garland.cpp:215:36: error: request for member 'size' in 'pals', which is of non-class type 'Palette [11]'
     paletteInd = secureRandom(pals.size());
                                    ^
*** [.pio\build\garland-wemos\src\garland.cpp.o] Error 1

In this case I need to define it's size using sizeof and can't using modern iteration. Do we really need it? Or I did something wrong?

Reply:
I may be wrong, but you can also do `std::distance(std::begin(pals), std::end(pals))` in place of `.size()` for basic arrays (which is why we could do `for (auto& x : pals) { ... }` stuff)
`std::array` also works, in case you don't mind typing out the size.

Reply:
Well, compiler don't let me use std::distance in constexpr function. But at least 
`constexpr size_t palsSize() { return sizeof(pals)/sizeof(pals[0]); }`
works well for me. Use pals[] with it.

Reply:
oops, yes. distance is compile-time with c++17, which is only available when building -git-base environments atm
(as well as simple `std::array something { 1,2,3,4,5 }` syntax...)